### PR TITLE
PIP-17:  provide DataBlockHeader and BlockAwareSegmentInputStream implementation

### PR DESCRIPTION
**Pull request by zhaijack**
### Motivation
- provide DataBlockHeader implementation;
- Implementation of InputStream(BlockAwareSegmentInputStream). when calls the is.read(), it return the following data:
1. Block header
2. Ledger entries starting from passed in entryId
3. Padding in the with byte 0xA

### Modifications

Add class DataBlockHeader and BlockAwareSegmentInputStream, 
Add test for them

### Result

unit test pass.

### CODE REVIEW

**Comment by zhaijack** [comment]
@ivankelly @sijie, Would you please help review it? Thanks.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/impl/BlockAwareSegmentInputStream.java [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+import org.slf4j.LoggerFactory;
+
+/**
+ *
+ * The BlockAwareSegmentInputStream for each cold storage data block.
+ * It contains a byte buffer, which contains all the content for this data block.
+ *      DataBlockHeader + entries(each with format[[entry_size -- int][entry_id -- long][entry_data]])
+ *
+ */
+public class BlockAwareSegmentInputStream extends ByteArrayInputStream {
```
Comment (ivankelly) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/impl/BlockAwareSegmentInputStream.java]: This is loading all the data for the block into memory. I specifically wanted to avoid this. And only load as needed. This shouldn't implement ByteArrayInputStream. It should implement InputStream. When read is called, it should first read out the block header, and then read from the ReadHandle on demand.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/DataBlockHeader.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+ * The data block header in code storage for each data block
+ *
+ * Currently, It is in format:
+ * [ magic_word -- int ][ block_len -- int ][ block_entry_count -- int ][ first_entry_id  -- long]
+ *
+ * with the size: 4 + 4 + 4 + 8 = 20 Bytes
+ */
+@Unstable
+@LimitedPrivate
+public interface DataBlockHeader {
```
Comment (ivankelly) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/DataBlockHeader.java]: Please split into separate PRs for DataBlockHeader and BlockAwareSegmentInputStream so they can be reviewed independently.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/DataBlockHeader.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.broker.s3offload;
+
+import java.io.IOException;
+import java.io.InputStream;
+import org.apache.bookkeeper.common.annotation.InterfaceAudience.LimitedPrivate;
```
Comment (ivankelly) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/DataBlockHeader.java]: These are internal to the broker, so they don't really need annotation like this.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/impl/BlockAwareSegmentInputStream.java (hunk 1 above)
Comment (zhaijack) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/impl/BlockAwareSegmentInputStream.java]: @ivankelly As you see, header is built when we process read from ReadHandle, By current design, we could not read header without reading all the data. One change we could do it put the header at the end of data block.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/DataBlockHeader.java (hunk 2 above)
Comment (zhaijack) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/DataBlockHeader.java]: Thanks, BlockAwareSegmentInputStream rely on  DataBlockHeader, so put them together. If you insist on, I will do the split, but the PR for BlockAwareSegmentInputStream will also be like this PR, which include all the  DataBlockHeader and BlockAwareSegmentInputStream.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/impl/BlockAwareSegmentInputStream.java (hunk 1 above)
Comment (ivankelly) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/impl/BlockAwareSegmentInputStream.java]: ah. I think block_len and block_entry_count come from an earlier design where we were pulling everything into memory before pushing the block to S3. They should be omitted. We don't need them on read in any case. Block length is fixed except in the last block. We store the count and length in the index in any case. 

On read, we can either use the index, or keep reading from the block until we find the padding pattern. The padding pattern, when read as 4 bytes, should yield a negative integer, to ensure it could never possibly be considered to be an entry_len. So, make the padding pattern something like 0xF012DEAD (i.e. F for most significant nibble).

It would be no harm to have a bit of extra space in the block header if we did want to add something later. Maybe make it 64bytes or so, padded with 0 after the magit word and first entry id.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/DataBlockHeader.java (hunk 2 above)
Comment (ivankelly) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/DataBlockHeader.java]: DataBlockHeader doesn't depend on BlockAwareSegmentInputStream though. I would really prefer PR to be broken into the smallest logical units for review. It helps the reviewer focus on a single issue, and allows the reviewer to see clearly what is being tested and what is not. Otherwise, people have a tendency to skip over the tests in review. It does mean there's be more roundtrips per line of code, since ironically smaller PRs tend to get more comments, but this should be fairly easy to manage with git worktree.

**Comment by zhaijack** [comment]
Need split and some change, close it and will open new ones

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/impl/BlockAwareSegmentInputStream.java (hunk 1 above)
Comment (zhaijack) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/impl/BlockAwareSegmentInputStream.java]: since block size is fixed except the last block, it could be get before reading all the entry, would like to keep block size in header, the  is useful to know the block size in BlockAwareSegmentInputStream.

Comment (ivankelly) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/impl/BlockAwareSegmentInputStream.java]: sure, that's fine by me. How does BlockAwareSegmentInputStream use the block size though?
